[bug] Nomad APM uses query that only collects metrics regarding a single client process

### Issue
The Nomad APM currently uses an API that only returns metrics for the client that was queried. This means allocations for task groups which are not running on the Nomad client which the Autoscaler queries will never be part of the scaling calculation. This affects all setups where the number on Nomad clients is greater than 1.

### Nomad Autoscaler version
master commit: `19f37d5877290a67ca32f81320ac5c923f855f6f`

### Config file
```
strategy "target-value" {
  driver = "target-value"
}
```

### Job file
```
job "example" {
  datacenters = ["dc1"]

  group "cache" {

    count = 1

    scaling {
      enabled = true
      min     = 0
      max     = 10

      policy {
        query  = "avg_cpu"

        strategy = {
          name = "target-value"

          config = {
            target = 20
          }
        }
      }
    }

    task "redis" {
      driver = "docker"

      config {
        image = "redis:3.2"
      }

      resources {
        cpu    = 500
        memory = 256

        network {
          mbits = 10
        }
      }
    }
  }
}

```

### Operating system and Environment details
Reproduced on MacOS

### Reproduction step
 * run the dev Nomad cluster available within the [repo](https://github.com/hashicorp/nomad/tree/master/dev/cluster) ensuring [allocation metrics](https://www.nomadproject.io/docs/configuration/telemetry/#telemetry-stanza) are enabled on the clients
* register the above job using `nomad run <jobfile>`
* run the autoscaler, ensuring the `-nomad-address` flag is set to the node that is not running the allocation and that the config file above is loaded
* observe log output [1] stating the metric cannot be found
* stop the running autoscaler, and start it again pointing it this time at the node that is running the allocation
* observe the log output [2] completing the scaling evaluation without error

Optionally using [JQ](https://stedolan.github.io/jq/) we can view the difference in the metric output:
 
 * not running the alloc:
```
$ curl -s localhost:7646/v1/metrics |jq '.Gauges[] |select(.Labels.task_group=="cache") |select(.Name=="nomad.client.allocs.cpu.total_percent")'
```
 
 * running the alloc:
```
$ curl -s localhost:8646/v1/metrics |jq '.Gauges[] |select(.Labels.task_group=="cache") |select(.Name=="nomad.client.allocs.cpu.total_percent")'
{
  "Labels": {
    "task": "redis",
    "namespace": "default",
    "host": "pathfinder.local",
    "job": "example",
    "task_group": "cache",
    "alloc_id": "614e0479-1fa3-9438-7138-518230c6f976"
  },
  "Name": "nomad.client.allocs.cpu.total_percent",
  "Value": 0.5465946197509766
}
```

[1]
```
2020-05-13T09:35:08.829+0200 [ERROR] agent: failed to query APM: policy_id=9ea8bc28-7051-bee0-f0dd-1afb6824d3a8 source=nomad-apm strategy=target-value target=nomad-target error="metric not found: avg_cpu/cache/example"
```

[2]
```
2020-05-13T09:46:10.707+0200 [INFO]  agent: querying APM: policy_id=9ea8bc28-7051-bee0-f0dd-1afb6824d3a8 source=nomad-apm strategy=target-value target=nomad-target
2020-05-13T09:46:10.707+0200 [DEBUG] internal_plugin.nomad-apm: expanded query: from=avg_cpu/cache/example to="&plugin.Query{Metric:"nomad.client.allocs.cpu.total_percent", Job:"example", Group:"cache", Operation:"avg"}"
2020-05-13T09:46:10.710+0200 [INFO]  agent: calculating new count: policy_id=9ea8bc28-7051-bee0-f0dd-1afb6824d3a8 source=nomad-apm strategy=target-value target=nomad-target
```

--- a/plugins/builtin/apm/nomad/plugin/metrics.go
+++ b/plugins/builtin/apm/nomad/plugin/metrics.go
@@ -4,107 +4,165 @@ import (
 	"fmt"
 	"math"
 	"strings"
+
+	"github.com/hashicorp/nomad/api"
 )
 
-type Metrics struct {
-	Counters []Counter
-	Gauges   []Gauge
-	Samples  []Sample
+// query is the plugins internal representation of a query and contains all the
+// information needed to perform a Nomad APM query.
+type query struct {
+	metric    string
+	job       string
+	group     string
+	operation string
 }
 
-type Counter struct {
-	Count  float64
-	Labels map[string]string
-	Max    float64
-	Mean   float64
-	Min    float64
-	Name   string
-	Rate   float64
-	Stddev float64
-	Sum    float64
-}
+const (
+	// queryOps are the supported operators.
+	queryOpSum = "sum"
+	queryOpAvg = "avg"
+	queryOpMax = "max"
+	queryOpMin = "min"
 
-type Gauge struct {
-	Labels map[string]string
-	Name   string
-	Value  float64
-}
-
-type Sample struct {
-	Counter
-}
-
-type Query struct {
-	Metric    string
-	Job       string
-	Group     string
-	Operation string
-}
+	// queryMetrics are the supported resources for querying.
+	queryMetricCPU = "cpu"
+	queryMetricMem = "memory"
+)
 
 func (a *APMPlugin) Query(q string) (float64, error) {
+
+	// Parse the query ensuring we have all information available to make all
+	// subsequent calls.
 	query, err := parseQuery(q)
 	if err != nil {
 		return 0, fmt.Errorf("failed to parse query: %v", err)
 	}
-
 	a.logger.Debug("expanded query", "from", q, "to", fmt.Sprintf("%# v", query))
 
-	var resp Metrics
-	_, err = a.client.Raw().Query("/v1/metrics", &resp, nil)
+	metrics, err := a.getTaskGroupResourceUsage(query)
 	if err != nil {
 		return 0, err
-	}
-
-	metrics := []Gauge{}
-	for _, g := range resp.Gauges {
-		if g.Name == query.Metric && g.Labels["job"] == query.Job && g.Labels["task_group"] == query.Group {
-			metrics = append(metrics, g)
-		}
 	}
 
 	if len(metrics) == 0 {
 		return 0, fmt.Errorf("metric not found: %s", q)
 	}
+	a.logger.Debug("metrics found", "num_data_points", len(metrics), "query", q)
 
-	var result float64
-	switch query.Operation {
-	case "sum":
-		for _, m := range metrics {
-			result += m.Value
+	return calculateResult(query.operation, metrics), nil
+}
+
+// getTaskGroupResourceUsage iterates the allocations within a job and
+// identifies those which meet the criteria for being part of the calculation.
+func (a *APMPlugin) getTaskGroupResourceUsage(query *query) ([]float64, error) {
+
+	// Grab the list of allocations assigned to the job in question.
+	allocs, _, err := a.client.Jobs().Allocations(query.job, false, nil)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get alloc listing for job: %v", err)
+	}
+
+	// The response is a list of data points from each allocation running in
+	// the task group.
+	var resp []float64
+
+	// Define a function that manages updating our response.
+	metricFunc := func(m *[]float64, ru *api.ResourceUsage) {}
+
+	// Depending on the desired metric, the function will append different data
+	// to the response. Using a function means we only have to perform the
+	// switch a single time, rather than on a per allocation basis.
+	switch query.metric {
+	case queryMetricCPU:
+		metricFunc = func(m *[]float64, ru *api.ResourceUsage) {
+			*m = append(*m, ru.CpuStats.Percent)
 		}
-	case "avg":
-		for _, m := range metrics {
-			result += m.Value
-		}
-		result /= float64(len(metrics))
-	case "max":
-		result = math.SmallestNonzeroFloat64
-		for _, m := range metrics {
-			if m.Value > result {
-				result = m.Value
-			}
-		}
-	case "min":
-		result = math.MaxFloat64
-		for _, m := range metrics {
-			if m.Value < result {
-				result = m.Value
-			}
+	case queryMetricMem:
+		metricFunc = func(m *[]float64, ru *api.ResourceUsage) {
+			*m = append(*m, float64(ru.MemoryStats.Usage))
 		}
 	}
 
-	return result, nil
+	for _, alloc := range allocs {
+
+		// If the allocation is not running, or is not part of the target task
+		// group then we should skip and move onto the next allocation.
+		if alloc.ClientStatus != api.AllocClientStatusRunning || alloc.TaskGroup != query.group {
+			continue
+		}
+
+		// Obtains the statistics for the task group allocation. If we get a
+		// single error during the iteration, we cannot reliably make a scaling
+		// calculation.
+		//
+		// When calling Stats an entire Allocation object is needed, but only
+		// the ID is used within the call. Further details:
+		// https://github.com/hashicorp/nomad/issues/7955
+		allocStats, err := a.client.Allocations().Stats(&api.Allocation{ID: alloc.ID}, nil)
+		if err != nil {
+			return nil, fmt.Errorf("failed to get alloc stats: %v", err)
+		}
+
+		// Be safe, be sensible.
+		if allocStats == nil {
+			continue
+		}
+
+		// Call the metric function to append the allocation resource metric to
+		// the response.
+		metricFunc(&resp, allocStats.ResourceUsage)
+	}
+
+	return resp, nil
 }
 
-func parseQuery(q string) (*Query, error) {
+// calculateResult determines the query result based on the metrics and
+// operation to perform.
+func calculateResult(op string, metrics []float64) float64 {
+
+	var result float64
+
+	switch op {
+	case queryOpSum:
+		for _, m := range metrics {
+			result += m
+		}
+	case queryOpAvg:
+		for _, m := range metrics {
+			result += m
+		}
+		result /= float64(len(metrics))
+	case queryOpMax:
+		result = math.SmallestNonzeroFloat64
+		for _, m := range metrics {
+			if m > result {
+				result = m
+			}
+		}
+	case queryOpMin:
+		result = math.MaxFloat64
+		for _, m := range metrics {
+			if m < result {
+				result = m
+			}
+		}
+	}
+	return result
+}
+
+// parseQuery takes the query string and transforms it into our internal query
+// representation. Parsing validates that the returned query is usable by all
+// subsequent calls but cannot ensure the job or group will actually be found
+// on the cluster.
+func parseQuery(q string) (*query, error) {
 	mainParts := strings.SplitN(q, "/", 3)
 	if len(mainParts) != 3 {
 		return nil, fmt.Errorf("expected <query>/<job>/group>, received %s", q)
 	}
 
-	query := &Query{
-		Group: mainParts[1],
-		Job:   mainParts[2],
+	query := &query{
+		group: mainParts[1],
+		job:   mainParts[2],
 	}
 
 	opMetricParts := strings.SplitN(mainParts[0], "_", 2)
@@ -116,19 +174,19 @@ func parseQuery(q string) (*Query, error) {
 	metric := opMetricParts[1]
 
 	switch metric {
-	case "cpu":
-		query.Metric = "nomad.client.allocs.cpu.total_percent"
-	case "memory":
-		query.Metric = "nomad.client.allocs.memory.usage"
+	case queryMetricCPU, queryMetricMem:
+		query.metric = metric
 	default:
-		query.Metric = metric
+		return nil, fmt.Errorf(`invalid metric %q, allowed values are %s or %s`,
+			metric, queryMetricCPU, queryMetricMem)
 	}
 
 	switch op {
-	case "sum", "avg", "min", "max":
-		query.Operation = op
+	case queryOpSum, queryOpAvg, queryOpMin, queryOpMax:
+		query.operation = op
 	default:
-		return nil, fmt.Errorf(`invalid operation "%s", allowed values are sum, avg, min or max`, op)
+		return nil, fmt.Errorf(`invalid operation %q, allowed values are %s, %s, %s or %s`,
+			op, queryOpSum, queryOpAvg, queryOpMin, queryOpMax)
 	}
 
 	return query, nil

--- a/plugins/builtin/apm/nomad/plugin/metrics_test.go
+++ b/plugins/builtin/apm/nomad/plugin/metrics_test.go
@@ -6,54 +6,84 @@ import (
 	"github.com/stretchr/testify/assert"
 )
 
+func Test_calculateResult(t *testing.T) {
+	testCases := []struct {
+		inputOp        string
+		inputMetrics   []float64
+		expectedOutput float64
+		name           string
+	}{
+		{
+			inputOp:        queryOpSum,
+			inputMetrics:   []float64{76.34, 13.13, 24.50},
+			expectedOutput: 113.97,
+			name:           "sum operation",
+		},
+		{
+			inputOp:        queryOpAvg,
+			inputMetrics:   []float64{76.34, 13.13, 24.50},
+			expectedOutput: 37.99,
+			name:           "avg operation",
+		},
+		{
+			inputOp:        queryOpMax,
+			inputMetrics:   []float64{76.34, 13.13, 24.50, 76.33},
+			expectedOutput: 76.34,
+			name:           "max operation",
+		},
+		{
+			inputOp:        queryOpMin,
+			inputMetrics:   []float64{76.34, 13.13, 24.50, 13.14},
+			expectedOutput: 13.13,
+			name:           "min operation",
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			actualOutput := calculateResult(tc.inputOp, tc.inputMetrics)
+			assert.Equal(t, tc.expectedOutput, actualOutput, tc.name)
+		})
+	}
+}
+
 func Test_parseQuery(t *testing.T) {
 	testCases := []struct {
 		name        string
 		input       string
-		expected    *Query
+		expected    *query
 		expectError bool
 	}{
 		{
 			name:  "avg_cpu",
 			input: "avg_cpu/group/job",
-			expected: &Query{
-				Metric:    "nomad.client.allocs.cpu.total_percent",
-				Job:       "job",
-				Group:     "group",
-				Operation: "avg",
+			expected: &query{
+				metric:    "cpu",
+				job:       "job",
+				group:     "group",
+				operation: "avg",
 			},
 			expectError: false,
 		},
 		{
 			name:  "avg_memory",
 			input: "avg_memory/group/job",
-			expected: &Query{
-				Metric:    "nomad.client.allocs.memory.usage",
-				Job:       "job",
-				Group:     "group",
-				Operation: "avg",
-			},
-			expectError: false,
-		},
-		{
-			name:  "arbritary metric",
-			input: "avg_nomad.client.allocs.cpu.total_percent/group/job",
-			expected: &Query{
-				Metric:    "nomad.client.allocs.cpu.total_percent",
-				Job:       "job",
-				Group:     "group",
-				Operation: "avg",
+			expected: &query{
+				metric:    "memory",
+				job:       "job",
+				group:     "group",
+				operation: "avg",
 			},
 			expectError: false,
 		},
 		{
 			name:  "job with fwd slashes",
 			input: "avg_cpu/group/my/super/job//",
-			expected: &Query{
-				Metric:    "nomad.client.allocs.cpu.total_percent",
-				Job:       "my/super/job//",
-				Group:     "group",
-				Operation: "avg",
+			expected: &query{
+				metric:    "cpu",
+				job:       "my/super/job//",
+				group:     "group",
+				operation: "avg",
 			},
 			expectError: false,
 		},

